packages/Pagination/src/Paginator/index.tsx

```diff
@@ -0,0 +1,94 @@
+/** headless component for pagination */
+import { useReducer } from 'react';
+import { fetchPageNumbers } from './utils';
+
+export interface PaginationOptions {
+  totalRecords: number;
+  pageSize: number;
+  pageNeighbors: number;
+}
+
+export interface PaginationState {
+  currentPage: number;
+  totalPages: number;
+  pagesToBeDisplayed: number[];
+  pageSize: number;
+  totalRecords: number;
+}
+
+export enum paginationActionTypes {
+  TO_PAGE = 'TO_PAGE'
+}
+
+export interface SwitchCurrentPageAction {
+  type: paginationActionTypes;
+  currentPage: number;
+  pagesToBeDisplayed: number[];
+}
+
+export function paginationReducer(state: PaginationState, action: SwitchCurrentPageAction) {
+  switch (action.type) {
+    case paginationActionTypes.TO_PAGE:
+      return {
+        ...state,
+        currentPage: action.currentPage,
+        pagesToBeDisplayed: action.pagesToBeDisplayed
+      };
+    default:
+      return state;
+  }
+}
+
+export const usePagination = (options: PaginationOptions) => {
+  const { totalRecords, pageSize, pageNeighbors } = options;
+  const totalPages = Math.ceil(totalRecords / pageSize);
+
+  // pageNeighbors can be in [0, 1, 2, 3, 4, 5]
+  const neighborPillsNum = Math.max(0, Math.min(pageNeighbors, 5));
+
+  const defaultPaginationState: PaginationState = {
+    currentPage: 1,
+    pageSize,
+    pagesToBeDisplayed: fetchPageNumbers(neighborPillsNum, totalPages, 0),
+    totalPages,
+    totalRecords
+  };
+
+  const [state, dispatch] = useReducer(paginationReducer, defaultPaginationState);
+
+  const nextPage = () =>
+    dispatch(changePageCreator(state.currentPage + 1, totalPages, pageNeighbors));
+  const firstPage = () => dispatch(changePageCreator(1, totalPages, pageNeighbors));
+  const lastPage = () => dispatch(changePageCreator(totalPages, totalPages, pageNeighbors));
+  const previousPage = () =>
+    dispatch(changePageCreator(state.currentPage - 1, totalPages, pageNeighbors));
+  const goToPage = (pageNumber: number) =>
+    dispatch(changePageCreator(pageNumber, totalPages, pageNeighbors));
+
+  const canPreviousPage = state.currentPage > 1;
+  const canNextPage = state.currentPage < totalPages;
+
+  return {
+    canNextPage,
+    canPreviousPage,
+    firstPage,
+    goToPage,
+    lastPage,
+    nextPage,
+    paginationState: state,
+    previousPage
+  };
+};
+
+const changePageCreator = (
+  page: number,
+  totalPages: number,
+  pageNeighbors: number
+): SwitchCurrentPageAction => {
+  const thisPage = Math.max(1, Math.min(page, totalPages));
+  return {
+    currentPage: thisPage,
+    pagesToBeDisplayed: fetchPageNumbers(pageNeighbors, totalPages, page),
+    type: paginationActionTypes.TO_PAGE
+  };
+};
```

Comment:
@moshthepitt What do you think about this approach, before i go all out on documenting it and testing it, added  2 trivial storybooks for this.

Reply:
@p-netm could you write a little on how you envision usage of this?

One thing we should think about is how to make this usable by many entities at the same time e.g. if I have a page listing people and another page listing locations, will this ducks module be able to handle them at the same time?

Reply:
this should handle any usecase that would need pagination, all it provides is an abstraction of essential logic when paginating, it leaves any other implementation details like the ui to the developer. Also i think this can now easily be used with routing, by delegating that part to the developer to decide how they want to do it.